SO-6477: rework scoring of lexical similarity suggester

cmark opened the pull request:
Jira: https://snowowl.atlassian.net/browse/SO-6477

cmark committed:
feat(suggest): rework how scores are computed in the new lexical...

...similarity suggester

cmark committed:
feat(core): add support for converting ComponentURI into...

...resource URI with ecl query part

cmark committed:
fix(suggest): print out received value without ecl query part

cmark committed:
feat(suggest): support separate exact and min terms in lexical similary.

...suggester

nagyo reviewed (approved):
LGTM

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/search/LexicalSimilarityTermFilter.java:
```diff
@@ -99,52 +116,60 @@ public LexicalSimilarityTermFilter withIgnoreStopwords() {
+		// for each received exactTerm, inject an exact match clause in a dismax query
+		var exactMatches = Expressions.bool();
+		exactTerms.forEach(exactTerm -> {
+			exactMatches.should(TermFilter.exact().term(exactTerm).caseSensitive(isCaseSensitive()).build().toExpression(field, textFieldSuffix, exactFieldSuffix, prefixFieldSuffix));
```
Lucene will complain after 1024 clauses by default. A length check or trimming the collection might be necessary to handle extreme inputs.

cmark replied:
I'll raise a TooCostly error if there are more than 100 exactTerms in the term filter.

cmark replied:
I added a bad request error in this case.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/uri/ComponentURI.java:
```diff
@@ -99,6 +99,11 @@ public final boolean isUnspecified() {
 	
+	@JsonIgnore
+	public final ResourceURIWithQuery asResourceURIWithQuery() {
+		return resourceUri().withQuery("ecl=" + identifier());
```
We also have a `withQueryPart` that re-assembles the ResourceURI side from its building blocks, which one should be used and when?

https://github.com/b2ihealthcare/snow-owl/blob/98b927ea6ce2fe160e78edf0065f3d539f6d73b2/core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/ResourceURI.java#L188-L191

https://github.com/b2ihealthcare/snow-owl/blob/98b927ea6ce2fe160e78edf0065f3d539f6d73b2/core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/ResourceURI.java#L295-L297

cmark replied:
Yeah, I saw that, will rework this part later.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/ResourceURIWithQuery.java:
```diff
@@ -64,8 +64,16 @@ public ResourceURIWithQuery(String uri) {
+		if (Strings.isNullOrEmpty(query)) {
+			this.query = query;
+			this.uri = resourceUri.toString();
+		} else if (query.startsWith(QUERY_PART_SEPARATOR)) {
+			this.query = query.substring(1);
+			this.uri = String.join("", resourceUri.toString(), query);
```
What was the reason for introducing this check?

cmark replied:
Double question mark in URIs.

cmark replied:
I'll add a test case for it.

cmark replied:
Done.

apeteri reviewed (approved):
LGTM with questions/suggestions below.

cmark committed:
test(core): add test cases to verify withQuery functionality

cmark committed:
fix(suggester): make sure no more than the first 100 terms...

...are considered for exact matching to avoid too many clauses error on
index level.

cmark committed:
feat(index): implement constant score expression to ES query conversion

cmark committed:
test(suggest): make sure we receive a 1.0 score in the test...

...as the given like text is a 1:1 match of a concept
